### PR TITLE
feat(sql): timestamp_floor() and timestamp_ceil() implementations

### LINKED ISSUE #1271: Round Up and Round Down timestamps.
**Is your feature request related to a problem? Please describe.**
I would like to have the ability to round up to nearest second/minute/hour, etc timestamps.

**Describe the solution you'd like**
To round down to nearest minute: `cast((from_timestamp / 60000000L) * 60000000L as timestamp)`
To round up to nearest minute: `dateadd('m', 1, cast((from_timestamp / 60000000L) * 60000000L as timestamp))`

**Describe alternatives you've considered**
None beside the solution above. They were provided by @ideoma (Thanks!)

**Additional context**
None.

### PR DESCRIPTION
**Pull request by bluestreak01**
closes #1271

### CODE REVIEW

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 92 / 93 (98.92%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/SqlKeywords.java|46|47|97.87%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/date/TimestampCeilFunctionFactory.java|20|20|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/date/TimestampFloorFunctionFactory.java|20|20|100.00%|
|:large_blue_circle:|io/questdb/std/datetime/microtime/Timestamps.java|6|6|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 92 / 93 (98.92%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/SqlKeywords.java|46|47|97.87%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/date/TimestampCeilFunctionFactory.java|20|20|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/date/TimestampFloorFunctionFactory.java|20|20|100.00%|
|:large_blue_circle:|io/questdb/std/datetime/microtime/Timestamps.java|6|6|100.00%|

File: core/src/main/java/io/questdb/std/datetime/microtime/Timestamps.java [hunk 1]
```diff
@@ -185,6 +185,14 @@ public static long ceilDD(long micros) {
                 ;
     }
 
+    public static long ceilHH(long micros) {
+        return floorHH(micros) + HOUR_MICROS - 1;
```
Comment (ideoma) [inline core/src/main/java/io/questdb/std/datetime/microtime/Timestamps.java:189]: I'd exepect it to be simply `floorHH(micros) + HOUR_MICROS` without -1

File: core/src/main/java/io/questdb/griffin/SqlKeywords.java [hunk 2]
```diff
@@ -637,6 +637,47 @@ public static boolean isHourKeyword(CharSequence tok) {
                 && (tok.charAt(i)) == 'r';
     }
 
+    public static boolean isMinuteKeyword(CharSequence tok) {
+        if (tok.length() != 6) {
+            return false;
+        }
+
+        int i = 0;
+        return (tok.charAt(i++)) == 'm'
```
Comment (ideoma) [inline core/src/main/java/io/questdb/griffin/SqlKeywords.java:646]: We have shortcuts in SAMPLE BY syntax for minute, hour, day etc, I'd prefer it to be supported here too

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 92 / 93 (98.92%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/SqlKeywords.java|46|47|97.87%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/date/TimestampCeilFunctionFactory.java|20|20|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/date/TimestampFloorFunctionFactory.java|20|20|100.00%|
|:large_blue_circle:|io/questdb/std/datetime/microtime/Timestamps.java|6|6|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 92 / 93 (98.92%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/SqlKeywords.java|46|47|97.87%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/date/TimestampCeilFunctionFactory.java|20|20|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/date/TimestampFloorFunctionFactory.java|20|20|100.00%|
|:large_blue_circle:|io/questdb/std/datetime/microtime/Timestamps.java|6|6|100.00%|

File: core/src/test/java/io/questdb/std/datetime/microtime/TimestampsTest.java [hunk 3]
```diff
@@ -95,6 +95,35 @@ public void testAddYearsPrevEpoch() throws Exception {
...
+        final long micros = TimestampFormatUtils.parseTimestamp("2021-09-09T22:44:56.108872Z");
+        TimestampFormatUtils.appendDateTime(sink, Timestamps.ceilMI(micros));
+        TestUtils.assertEquals("2021-09-09T22:44:59.999Z", sink);
+    }
+
+    @Test
+    public void testCeilMS() throws Exception {
+        final long micros = TimestampFormatUtils.parseTimestamp("2021-09-09T22:44:56.108872Z");
+        TimestampFormatUtils.appendDateTime(sink, Timestamps.ceilMS(micros));
+        // TODO: is this correct or should yield .109Z ?
```
[the remark this answers is not in the document]
Comment (bluestreak01) [reply core/src/test/java/io/questdb/std/datetime/microtime/TimestampsTest.java]: yes, done

**Review by ideoma — DISMISSED** [review]

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 41 / 41 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/date/TimestampCeilFunctionFactory.java|14|14|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/date/TimestampFloorFunctionFactory.java|13|13|100.00%|
|:large_blue_circle:|io/questdb/std/datetime/microtime/Timestamps.java|8|8|100.00%|
|:large_blue_circle:|io/questdb/cairo/PartitionBy.java|5|5|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 41 / 41 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/date/TimestampCeilFunctionFactory.java|14|14|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/date/TimestampFloorFunctionFactory.java|13|13|100.00%|
|:large_blue_circle:|io/questdb/std/datetime/microtime/Timestamps.java|8|8|100.00%|
|:large_blue_circle:|io/questdb/cairo/PartitionBy.java|5|5|100.00%|

**Review by marregui — APPROVED** [review]
LGTM

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 41 / 41 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/date/TimestampCeilFunctionFactory.java|14|14|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/date/TimestampFloorFunctionFactory.java|13|13|100.00%|
|:large_blue_circle:|io/questdb/std/datetime/microtime/Timestamps.java|8|8|100.00%|
|:large_blue_circle:|io/questdb/cairo/PartitionBy.java|5|5|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 41 / 41 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/date/TimestampCeilFunctionFactory.java|14|14|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/date/TimestampFloorFunctionFactory.java|13|13|100.00%|
|:large_blue_circle:|io/questdb/std/datetime/microtime/Timestamps.java|8|8|100.00%|
|:large_blue_circle:|io/questdb/cairo/PartitionBy.java|5|5|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 41 / 41 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/date/TimestampCeilFunctionFactory.java|14|14|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/date/TimestampFloorFunctionFactory.java|13|13|100.00%|
|:large_blue_circle:|io/questdb/std/datetime/microtime/Timestamps.java|8|8|100.00%|
|:large_blue_circle:|io/questdb/cairo/PartitionBy.java|5|5|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 41 / 41 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/date/TimestampCeilFunctionFactory.java|14|14|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/date/TimestampFloorFunctionFactory.java|13|13|100.00%|
|:large_blue_circle:|io/questdb/std/datetime/microtime/Timestamps.java|8|8|100.00%|
|:large_blue_circle:|io/questdb/cairo/PartitionBy.java|5|5|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 41 / 41 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/date/TimestampCeilFunctionFactory.java|14|14|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/date/TimestampFloorFunctionFactory.java|13|13|100.00%|
|:large_blue_circle:|io/questdb/std/datetime/microtime/Timestamps.java|8|8|100.00%|
|:large_blue_circle:|io/questdb/cairo/PartitionBy.java|5|5|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 41 / 41 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/date/TimestampCeilFunctionFactory.java|14|14|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/date/TimestampFloorFunctionFactory.java|13|13|100.00%|
|:large_blue_circle:|io/questdb/std/datetime/microtime/Timestamps.java|8|8|100.00%|
|:large_blue_circle:|io/questdb/cairo/PartitionBy.java|5|5|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 41 / 41 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/date/TimestampCeilFunctionFactory.java|14|14|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/date/TimestampFloorFunctionFactory.java|13|13|100.00%|
|:large_blue_circle:|io/questdb/std/datetime/microtime/Timestamps.java|8|8|100.00%|
|:large_blue_circle:|io/questdb/cairo/PartitionBy.java|5|5|100.00%|

**Review by ideoma — APPROVED** [review]

**Pull request merged**